bindings/python/unicorn/unicorn.py

```diff
@@ -1,492 +1,303 @@
-        reg = ctypes.c_uint64(value)
-
-    status = reg_write_func(reg_id, ctypes.byref(reg))
-    if status != uc.UC_ERR_OK:
-        raise UcError(status)
-
-    return
-
-def _catch_hook_exception(func):
-    @wraps(func)
-    def wrapper(self, *args, **kwargs):
-        """Catches exceptions raised in hook functions.
+    major = uc.UC_API_MAJOR
+    minor = uc.UC_API_MINOR
 
-        If an exception is raised, it is saved to the Uc object and a call to stop
-        emulation is issued.
-        """
-        try:
-            return func(self, *args, **kwargs)
-        except Exception as e:
-            # If multiple hooks raise exceptions, just use the first one
-            if self._hook_exception is None:
-                self._hook_exception = e
+    combined = (major << 8) + minor
 
-            self.emu_stop()
+    return (major, minor, combined)
+
+
+def uc_arch_supported(atype: int) -> bool:
+    """Check whether Unicorn library supports a particular arch.
+    """
+
+    return bool(uclib.uc_arch_supported(atype))
 
-    return wrapper
+
+def debug() -> str:
+    """Get verbose verion string.
+    """
+
+    archs = (
+        ('arm',     uc.UC_ARCH_ARM),
+        ('arm64',   uc.UC_ARCH_ARM64),
+        ('mips',    uc.UC_ARCH_MIPS),
+        ('x86',     uc.UC_ARCH_X86),
+        ('ppc',     uc.UC_ARCH_PPC),
+        ('sparc',   uc.UC_ARCH_SPARC),
+        ('m68k',    uc.UC_ARCH_M68K),
+        ('riscv',   uc.UC_ARCH_RISCV),
+        ('s390x',   uc.UC_ARCH_S390X),
+        ('tricore', uc.UC_ARCH_TRICORE)
+    )
+
+    all_archs = ''.join(f'-{name}' for name, atype in archs if uc_arch_supported(atype))
+    lib_maj, lib_min, _ = uc_version()
+    bnd_maj, bnd_min, _ = version_bind()
+
+    return f'python-{all_archs}-c{lib_maj}.{lib_min}-b{bnd_maj}.{bnd_min}'
 
 
-class uc_arm_cp_reg(ctypes.Structure):
-    """ARM coprocessors registers for instructions MRC, MCR, MRRC, MCRR"""
-    _fields_ = [
-        ("cp", ctypes.c_uint32),
-        ("is64", ctypes.c_uint32),
-        ("sec", ctypes.c_uint32),
-        ("crn", ctypes.c_uint32),
-        ("crm", ctypes.c_uint32),
-        ("opc1", ctypes.c_uint32),
-        ("opc2", ctypes.c_uint32),
-        ("val", ctypes.c_uint64)
-    ]
-
-class uc_arm64_cp_reg(ctypes.Structure):
-    """ARM64 coprocessors registers for instructions MRS, MSR"""
-    _fields_ = [
-        ("crn", ctypes.c_uint32),
-        ("crm", ctypes.c_uint32),
-        ("op0", ctypes.c_uint32),
-        ("op1", ctypes.c_uint32),
-        ("op2", ctypes.c_uint32),
-        ("val", ctypes.c_uint64)
-    ]
-
-class uc_x86_mmr(ctypes.Structure):
-    """Memory-Management Register for instructions IDTR, GDTR, LDTR, TR."""
-    _fields_ = [
-        ("selector", ctypes.c_uint16),  # not used by GDTR and IDTR
-        ("base", ctypes.c_uint64),      # handle 32 or 64 bit CPUs
-        ("limit", ctypes.c_uint32),
-        ("flags", ctypes.c_uint32),     # not used by GDTR and IDTR
-    ]
-
-class uc_x86_msr(ctypes.Structure):
```

Comment:
I love your new design of the registers read/write per arch, but it is a bit aggressive to remove these classes since they might be really used somewhere, though unexpected.

Reply:
These classes represent registers in the way Uc sees them, so I don't think they were meant to be used externally.
As stated above, this is another step towards a cleaner namespace.